doc/usecases/core/UC_keyset_lookup_direct.md

```diff
@@ -14,9 +14,9 @@
 - **Main success scenario:**
   - Caller asks Core to look up `Key` by non-cascading name in `KeySet`
   - Core searches for `Key` with same name in `KeySet`
-  - If a matching `Key` is found, Core returns a `Key *` to it.
-    The name of the `Key` will be read-only, otherwise it is modifiable.
-  - Otherwise, Core returns `NULL`
+  - If a matching `Key` is found, Core returns its index.
+  - Otherwise, Core returns a negative value, which indicates "not found".
+    Additionally, the negative value can be converted (with a well-defined formula) into the index at which the searched `Key` would be inserted.
```

Comment:
What is this "well-defined formula"? I assume it is something like `(returned value + 1) * (-1)`?

Reply:
Yes, that's currently the plan, I just didn't want to specify it here, because for the use case it doesn't matter what the exact formula is. It's even a bit much to specify a negative value.

Maybe it would be better to say:

```markdown
- Otherwise, Core returns a value, which indicates "not found".
  Additionally, the index at which the searched `Key` would be inserted can be determined by Caller without any additional calls to Core.
```

In theory the API could look like this:

```c
typedef struct {
    size_t index : (sizeof(size_t)*CHAR_BIT-1);
    bool found : 1;
} LookupResult;

LookupResult ksLookup(KeySet * ks, Key * k);
```

Or this:

```c
// returns true, if found and false if not found
// *index is set to index of key if found, or index where key would be inserted if not found
bool ksLookup (KeySet * ks, Key * k, size_t * index); 
```

Reply:
If we already know that we want negative values, it is easier to write it down. No need to keep a UC abstract if within Elektra it needs to be specified anyway. If an alternative implementation could do it differently, it would be a different story.

Reply:
> If we already know that we want negative values

Do we? I actually think the `LookupResult` from above is an equally good solution. [The struct uses 8 bytes](https://godbolt.org/z/ahaWxGvcK) just like a `size_t`/`ssize_t` would. It might even be better, because it's much clearer how to interpret the result.

> No need to keep a UC abstract if within Elektra it needs to be specified anyway.

IMO there has to be a balance. Making it unnecessarily abstract and harder to read is bad. But making it overly specific and thereby limiting what can be done on the decision/implementation level without changing the use case is also bad.

Reply:
Ok, then let us keep it open in the use case.